### ISSUE TITLE
Validate class arguments

### ISSUE THREAD

**Issue by carmocca** [omni-us/jsonargparse#14]
Hi!

In my code, I validate the inputs to the parser with the `add_argument` type, e.g.

```python
import argparse


class PositiveNumber:
    def __init__(self, type):
        self.type = type

    def __call__(self, v):
        v = self.type(v)
        if v < 0:
            raise argparse.ArgumentTypeError("Must be positive")
        return v

parser = argparse.ArgumentParser()
parser.add_argument("--learning_rate", type=PositiveNumber(float))
args = parser.parse_args(["--learning_rate=-1"])
"""
usage: scratch_2.py [-h] [--learning_rate LEARNING_RATE]
scratch_2.py: error: argument --learning_rate: Must be positive
"""
```

However, it is not currently possible with `add_class_arguments`:

```python
import argparse
from dataclasses import dataclass

import jsonargparse


class PositiveNumber:
    def __init__(self, type):
        self.type = type

    def __call__(self, v):
        v = self.type(v)
        if v < 0:
            raise argparse.ArgumentTypeError("Must be positive")
        return v

@dataclass
class Optimizer:
    """
    Optimizer options

    Args:
        learning_rate: Learning rate (must be > 0)
    """
    learning_rate: PositiveNumber(float) = 0.001

# AssertionError is not raised. __call__ is not called
# expectedly, python does not validate
x = Optimizer(learning_rate=-1)

parser = jsonargparse.ArgumentParser()
parser.add_class_arguments(Optimizer, "optimizer")
args = parser.parse_args(["--optimizer.learning_rate=-1"])
# I believe this is a bug, fails to parse at all
# works with learning_rate: float = 0.001
"""
usage: scratch_1.py [-h] [--print-config]
scratch_1.py: error: Unrecognized arguments: --optimizer.learning_rate=-1
"""
```

Would it be possible for jsonargparse to run `__call__` on the types of each argument?

**Comment by carmocca** [comment]
Maybe what I am trying to do with `add_class_arguments` is not well suited for it. Are there any alternatives to organize arguments into their own groups but still validate them?

**Comment by carmocca** [comment]
I thought about this some more and found a workaround:

```python
def check_positive(v):
    if v < 0:
        raise argparse.ArgumentTypeError("Must be positive")

@dataclass
class Optimizer:
    """
    Optimizer options

    Args:
        learning_rate: Learning rate (must be > 0)
    """
    learning_rate: float = 0.001

    def __post_init__(self):
        check_positive(self.learning_rate)

parser = jsonargparse.ArgumentParser(parse_as_dict=True)
parser.add_class_arguments(Optimizer, "optimizer")

# ArgumentTypeError is not raised because
# the dataclass is not created
args = parser.parse_args(["--optimizer.learning_rate=-1"])

# It is raised here
args["optimizer"] = Optimizer(**args["optimizer"])
```

It has the drawback of having to convert the args dict/Namespace back into the dataclass manually so `__post_init__` is run.

Maybe jsonargparse could provide a `add_dataclass_arguments` method so the dataclass is automatically created and returned instead of a dict/Namespace

```python
parser.add_dataclass_arguments(Optimizer, "optimizer")
args = parser.parse_args()
assert isinstance(args["optimizer"], Optimizer) # if parse_as_dict
assert isinstance(args.optimizer, Optimizer) # if not parse_as_dict
```

What do you think?

**Comment by mauvilsa** [comment]
Interesting, i didn't know about `__post_init__`.

Currently arguments with unknown types are ignored and not added to the parser. However, I do see a nice new feature that could be added to `add_dataclass_arguments` to support this. Taking from your initial example the idea would be

```python
from typing import Type

PositiveFloatType = Type[PositiveNumber(float)]

@dataclass
class Optimizer:
    learning_rate: PositiveFloatType = 0.001

parser = jsonargparse.ArgumentParser()
parser.add_class_arguments(Optimizer, "optimizer", custom_types=[PositiveFloatType])
```

It would also make sense to have in jsonargparse a class `RestrictedNumberType` to easily define these types without needing to write a new class each time. It would be kind of a modification of ActionOperators that is already there https://omni-us.github.io/jsonargparse/#comparison-operators.

A downside I see is that probably `Optimizer` should again check that the number is positive because people would might want to use the class without a parser. This is redundant. Also we would need to see what mypy says about this. If you do `Optimizer(learning_rate=0.001)` it could be an error since it is not known that `0.001` is of type `PositiveFloatType`.

**Comment by carmocca** [comment]
What about a dataclass with many fields? Would you have to pass the list of all the custom_types used? Does order matter?

> `Optimizer` should again check that the number is positive because people would might want to use the class without a parser. This is redundant

Given this downside, I don't see it as a good approach. Code would end up duplicated and It would force using the ActionOperators inside the dataclass' `__post_init__` to have the exact same behaviour with or without a parser.

I feel what I proposed at the end of my last message is the cleanest approach. Maybe instead of adding the `add_dataclass_arguments` method, `add_class_arguments` could have an option `as_class: bool = False` to create the class and use it after `parse_args` is called

**Comment by mauvilsa** [comment]
Looking at errors that mypy was giving, I have realized that the proper way to define a "positive float type" would be

```python
class PositiveFloat(float):
    def __new__(self, v):
        v = float(v)
        if v < 0:
            raise ValueError("Must be positive")
        return super().__new__(self, v)

@dataclass
class Optimizer:
    learning_rate: PositiveFloat = PositiveFloat(0.001)
```

I think with this everything could be done in a clean way. Though it still requires to adapt jsonargparse to support it. I will think about the full details and let you know.

**Comment by mauvilsa** [comment]
Have a look at [typing_tests.py](https://github.com/omni-us/jsonargparse/blob/master/jsonargparse_tests/typing_tests.py) and [signatures_tests.py#L238-L247](https://github.com/omni-us/jsonargparse/blob/master/jsonargparse_tests/signatures_tests.py#L238-L247). Works nicely.

As a side note, the following would work. The only caveat is if someone wants to run mypy and not get an error, then the casting is needed.

```python
@dataclass
class Optimizer:
    learning_rate: PositiveFloat = 0.001
```

**Comment by carmocca** [comment]
Im trying out master right now. Looks sweet.

Have you thought about overridding `__str__` for the `RestrictedNumberType` class?

```python
def __str__(self):
    return f"{super().__repr__()} ({self._expression})"
```

Which would produce a help message like:

```
Optimizer options:
  --optimizer.learning_rate LEARNING_RATE
                        Learning rate (must be > 0) (default: 0.001 (v>0))
```

**Comment by carmocca** [comment]
This also gave me the idea of supporting `choices`:

```python
def choices_type(name, base_type, choices):
    choices = set(choices)

    class ChoicesType(base_type):
        _choices = choices
        _type = base_type

        def __new__(self, v):
            v = self._type(v)
            if v not in self._choices:
                raise ValueError(f'invalid value, {repr(v)} does not conform to choices {self._choices}')
            return super().__new__(self, v)

    ChoicesType.__name__ = name
    return ChoicesType

OptimChoices = choices_type("OptimChoices", str, ["Adam", "SGD", "AdaGrad"])

@dataclass
class Optimizer:
    """
    Optimizer options

    Args:
        algorithm: Optimizer algorithm
    """
    algorithm: OptimChoices = "Adam"


parser = jsonargparse.ArgumentParser()
parser.add_class_arguments(Optimizer)
```

with a docstring like

```
Optimizer options:
  --algorithm ALGORITHM
                        Optimizer algorithm (default: Adam) (choices: {Adam,SGD,AdaGrad})
```

**Comment by mauvilsa** [comment]
> This also gave me the idea of supporting `choices`:

This is already supported with `Enum`.

**Comment by mauvilsa** [comment]
> Have you thought about overridding `__str__` for the `RestrictedNumberType` class?

I did think a bit about this but I am still not sure what to do. Probably the best option is the help formatter, but extending the existing formatter, not a new one.

**Comment by carmocca** [comment]
A few more things:

- Is `List[RestrictedNumberType]` supported?

```python
from dataclasses import dataclass
from typing import List
import jsonargparse
from jsonargparse.typing import PositiveInt


@dataclass
class Foo:
    values: List[PositiveInt] = (1, 3, 5)

parser = jsonargparse.ArgumentParser()
parser.add_class_arguments(Foo)

args = parser.parse_args(["--values=[5,3,1]"], with_meta=False)
# error: Unrecognized arguments: --values=[5,3,1]
```
- How are tuples passed via command line args? `--values=(5,3,1)` does not work

- It is important that the typing classes can be pickled:

```python
import pickle
from jsonargparse.typing import PositiveInt

with open("/tmp/kk", "w") as f:
    pickle.dump(PositiveInt, f)
# AttributeError: Can't pickle local object 'restricted_number_type.<locals>.RestrictedNumberType'
```
Which means that having the `RestrictedNumberType` class definition inside the `restricted_number_type` function is not allowed.
This is because multiprocessing requires all spawned processes to be entirely pickleable.

**Comment by mauvilsa** [comment]
>     * Is `List[RestrictedNumberType]` supported?

Not yet.

>     * How are tuples passed via command line args? `--values=(5,3,1)` does not work

They are passed as lists `--values=[5,3,1]` but after parsing it will be a `tuple`. The limitations of this is explained in the readme.

>     * It is important that the typing classes can be pickled:

![person-facepalming_1f926](https://user-images.githubusercontent.com/5780272/98387797-28de5a80-2052-11eb-95e3-c20b5b0ffc91.png)

It would be a shame not being able to support the dynamic types. But supporting multiprocessing is important. I will look to see if it is possible to make it work.

**Comment by mauvilsa** [comment]
Making the dynamic types pickle wasn't as hard as I thought. Could you please test to verify that it works correctly with multiprocessing?

**Comment by carmocca** [comment]
> Could you please test to verify that it works correctly with multiprocessing?

It does!

**Comment by mauvilsa** [comment]
> > Have you thought about overridding `__str__` for the `RestrictedNumberType` class?
> 
> I did think a bit about this but I am still not sure what to do. Probably the best option is the help formatter, but extending the existing formatter, not a new one.

Now the help includes the type. Not just for restricted numbers but for all types. For example:

```
usage: types_help.py [-h] [--print-config] [--data.arg1 ARG1] [--data.arg2 ARG2] [--data.arg3 ARG3] [--data.arg4 ARG4] [--data.arg5 {X,Y,Z}]
                     [--data.arg6 {X,Y,Z,null}] [--data.arg7 ARG7] [--data.arg8 ARG8] [--data.arg9 ARG9]

optional arguments:
  -h, --help            show this help message and exit
  --print-config        print configuration and exit

DataClass:
  --data.arg1 ARG1      arg1 description (type: float v>0, default: 0.001)
  --data.arg2 ARG2      arg2 description (type: int, default: 5)
  --data.arg3 ARG3      arg3 description (type: List[float], default: [1.1, 1.2])
  --data.arg4 ARG4      arg4 description (type: Union[bool, null], default: null)
  --data.arg5 {X,Y,Z}   arg5 description (type: MyEnum, default: Y)
  --data.arg6 {X,Y,Z,null}
                        arg6 description (type: Union[MyEnum, null], default: Z)
  --data.arg7 ARG7      arg7 description (type: float v>=0 and v<=1, default: 0.5)
  --data.arg8 ARG8      arg8 description (type: int v>=10 and v<=20, default: 12)
  --data.arg9 ARG9      arg9 description (type: Dict[str, float], default: {'a': 0.1, 'b': 0.2})
```
